core/domain/opportunity_services.py

```diff
@@ -407,6 +407,147 @@ def update_opportunities_with_new_topic_name(topic_id, topic_name):
+    """Updates the skill_description of the SkillOpportunityModel with
+    new_description.
+
+    Args:
+        skill_id: str. The corresponding skill_id of the opportunity.
+        new_description: str. The new skill_description.
+    """
+    skill_opportunity_model = (
+        opportunity_models.SkillOpportunityModel.get(skill_id))
+    skill_opportunity_model.skill_description = new_description
+    skill_opportunity_model.put()
+
+
+def delete_skill_opportunities(skill_id):
+    """Deletes the SkillOpportunityModel corresponding to the supplied skill_id.
+
+    Args:
+        skill_id: str. The skill_id corresponding to the to-be-deleted
+            SkillOpportunityModel.
+    """
+    skill_opportunity_model = (
+        opportunity_models.SkillOpportunityModel.get(skill_id))
+    opportunity_models.SkillOpportunityModel.delete(skill_opportunity_model)
+
+
+def update_skill_opportunity_question_counts(skill_ids, delta):
+    """Increments question_count(s) of SkillOpportunityModel(s) with
+    corresponding skill_ids.
+
+    Args:
+        skill_ids: list(str). A list of skill_ids corresponding to
+            SkillOpportunityModel(s).
+        delta: int. The delta for which to update each question_count.
+    """
+    updated_models = _update_skill_opportunity_question_count(skill_ids, delta)
+    opportunity_models.SkillOpportunityModel.put_multi(updated_models)
+
+
+def compute_question_linked_skill_ids_update(old_skill_ids, new_skill_ids):
+    """Updates question_count(s) of SkillOpportunityModel(s) by the set
+    difference of new_skill_ids - old_skill_ids.
+
+    Args:
+        old_skill_ids: list(str). A list of old skill_id(s).
+        new_skill_ids: list(str). A list of new skill_id(s).
+    """
+    old_skill_ids_set = set(old_skill_ids)
+    new_skill_ids_set = set(new_skill_ids)
+    added = new_skill_ids_set - old_skill_ids_set
+    removed = old_skill_ids_set - new_skill_ids_set
+    updated_models = []
+    updated_models.extend(_update_skill_opportunity_question_count(added, 1))
+    updated_models.extend(_update_skill_opportunity_question_count(removed, -1))
+    opportunity_models.SkillOpportunityModel.put_multi(updated_models)
+
+
+def _update_skill_opportunity_question_count(skill_ids, delta):
+    """Returns a list of SkillOpportunityModel(s) with corresponding skill_ids
+    with question_count(s) updated by delta.
+
+    Args:
+        skill_ids: iterable(str). The skill_ids of the SkillOpportunityModel(s).
+        delta: int. The delta for which to update each question_count (can be
+            negative).
+
+    Returns:
+        updated_models. list(SkillOpportunityModel) The updated
+            SkillOpportunityModel(s).
+    """
+    updated_models = []
+    for skill_id in skill_ids:
+        skill_opportunity_model = (
+            opportunity_models.SkillOpportunityModel.get(skill_id))
+        if skill_opportunity_model:
+            skill_opportunity_model.question_count += delta
```

Comment:
Actually I just realized this isn't really the way to do it. We never change the model directly or do logic on it -- we do logic on the domain object, and just do pure gets/puts that transform the domain object from/to a model. All the logical operations and changes should happen on the domain object, and you can define methods on the domain object to help with that.

Could you please refactor accordingly? The only areas where an opportunity model should appear are when you're retrieving it from the datastore (to immediately convert it into a domain object), and when you're saving a domain object (in which case you'd transform it to a storage model and then do a put() on it without further operations in between).

Reply:
Done.